proposals/fleet-multiple-abm-vpp.md

```diff
@@ -0,0 +1,146 @@
+### Multiple ABM/VPP tokens
+
+- Figma: https://www.figma.com/design/j2M1heOh8eZD6LcUJks6HE/%239956-Add-multiple-Apple-Business-Manager-and-Volume-Purchasing-Program-connections?node-id=2-130&t=3vPRYgndVgzoD8e4-0
+- Issue: https://github.com/fleetdm/fleet/issues/9956
+
+- Constraints/bussines logic:
+    - VPP tokens can be assigned to: "All teams", "No team" or a team. Default is "All teams"
+    - ABM tokens can be assigned to: "No team" or a team. Default is "No team"
+    - Multiple ABM/VPP tokens per team.
+    - Same private key can be used for multiple tokens.
+
+### ABM: gameplan overview
+
+The ADE workflow is summarized here
+
+https://github.com/fleetdm/fleet/blob/0a2a48b6d89ab9e428509eb918724c60c245db60/tools/mdm/apple/glossary-and-protocols.md?plain=1#L216-L235
+
+Currently, Fleet manages a single token, and it does the process described there in a cron job every 30 seconds.
+
+To support multiple tokens, we need to:
+
+- On each cron run, do the process for each token that was uploaded.
+- For each host ingested via ABM, keep track of the token that was used to ingest it.
+- Every time we need to make an ABM call to assign an ADE profile, choose the right token to use for the host.
+
+### VPP: gameplan overview
+
+The VPP token is used to:
+
+- Get a list of available apps for a team
+- Assign an app to a host
+
+To support multiple tokens, we need to:
+
+- Add a form to retrieve all tokens available for a team or "no team" this is `(team OR no team) ∪ all teams`.
+- To get a list of available apps: loop over each token, get and dedupe the available apps.
+- To assign an app to a host: loop over each token until we find the first one with available licenses, use it to assign the app to the host.
+- Track the token used to assign the app to the host.
+
+
+### Database migrations
+
+- Ticket: https://github.com/fleetdm/fleet/issues/21176
+
+#### Migrations
+
+**Tracking team associations for ABM/VPP tokens**
+
+Many-to-many relationship to be tracked in a join table:
+
+```sql
+CREATE TABLE mdm_asset_assignments (
+    id INT PRIMARY KEY,
+    asset_id INT NOT NULL,
+    team_id INT NULL, -- NULL if the asset is assigned to "no team" or "all teams"
+    target_platform varchar(10) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL, -- if NULL, asset is not platform specific
+    scope ENUM('all_teams', 'team', 'no_team') NOT NULL,
+    FOREIGN KEY (asset_id) REFERENCES mdm_config_assets(id),
+    FOREIGN KEY (team_id) REFERENCES teams(id),
+);
+```
+
+**Adding new tokens to mdm_config_assets**
+
+
+Allow multiple assets with the same name:
```

Comment:
We'll need to track uniqueness of ABM and VPP tokens. For ABM, it is the organization name, and for VPP tokens, it is the location: https://www.figma.com/design/j2M1heOh8eZD6LcUJks6HE/%239956-Add-multiple-Apple-Business-Manager-and-Volume-Purchasing-Program-connections?node-id=5316-3424&t=anBMV3xXIbMrBv64-0

Not sure if `mdm_config_assets` is the place to do this or if we need a different table...  We also need to be able to return that "Organization name" and "Location" (for VPP) associated with a token, though maybe that information is available as part of the decoded/decrypted token? But even if so, that might mean to much work to do on every request to display this information, we should probably store that in cleartext (we kinda have to to get the uniqueness check anyway).

Reply:
Related, on "renew" we need to replace the token itself but otherwise keep the configuration for that token unchanged (associated teams and all).

Reply:
> At this point, should we consider separate tables for each instead of a single mdm_config_assets?

Yeah that makes sense to me, at that point we're kinda fighting to fit it in the `mdm_config_assets`. We could then have proper org name, location and renew date columns along with the (encrypted, I think?) token itself, and the relevant keys/uniqueness constraints.

Reply:
And a `renew` API call could be an update with the `id` of that table's row, so we know exactly which token to update/renew.

Reply:
@mna with the changes discussed at stand-up, should we keep the join table? an alternative approach could be:

1. add two tables `vpp_tokens` and `abm_tokens`
2. add two nulable foreign key columns to `teams`, one for each table.

Reply:
@roperzh I agree with the alternative approach you suggest. If we move to many-to-many, the required migration would be relatively straightforward, and for now it keeps things simple enough. Though I think we still need the "no team" and "all teams" distinction?